src/dbnode/storage/index.go

```diff
@@ -1554,52 +1605,73 @@ func (i *nsIndex) queryWithSpan(
-			exhaustive: true,
-		}
-		wg            sync.WaitGroup
-		totalWaitTime time.Duration
+		state   = asyncQueryExecState{}
+		wg      sync.WaitGroup
+		permits = i.permitsManager.NewPermits(ctx)
 	)
 
-	for _, block := range blocks {
-		// Capture block for async query execution below.
-		block := block
-
-		// We're looping through all the blocks that we need to query and kicking
-		// off parallel queries which are bounded by the queryWorkersPool's maximum
-		// concurrency. This means that it's possible at this point that we've
-		// completed querying one or more blocks and already exhausted the maximum
-		// number of results that we're allowed to return. If thats the case, there
-		// is no value in kicking off more parallel queries, so we break out of
-		// the loop.
-		seriesCount := results.Size()
-		docsCount := results.TotalDocsCount()
-		alreadyExceededLimit := opts.SeriesLimitExceeded(seriesCount) || opts.DocsLimitExceeded(docsCount)
-		if alreadyExceededLimit {
-			state.Lock()
-			state.exhaustive = false
-			state.Unlock()
-			// Break out if already not exhaustive.
-			break
+	blockIters := make([]*blockIter, len(blocks))
+	for b, block := range blocks {
+		iter, err := newBlockIterFn(ctx, block, query, results)
+		if err != nil {
+			return false, err
+		}
+		blockIters[b] = &blockIter{
+			iter:       iter,
+			iterCloser: x.NewSafeCloser(iter),
+			block:      block,
+		}
+	}
+
+	defer func() {
+		for _, iter := range blockIters {
+			// safe to call Close multiple times, so it's fine to eagerly close in the loop below and here.
+			_ = iter.iterCloser.Close()
 		}
+	}()
+
+	// used by each blockIter to break the iteration loop if the query can't continue processing.
+	breakIter := func() bool {
+		return opts.LimitsExceeded(results.Size(), results.TotalDocsCount()) || state.hasErr()
+	}
 
-		// Calculate time spent waiting for a worker
+	// We're looping through all the blocks that we need to query and kicking
+	// off parallel queries which are bounded by the permits maximum
+	// concurrency. It's possible at this point that we've completed querying one or more blocks and already exhausted
+	// the maximum number of results that we're allowed to return. If thats the case, there is no value in kicking off
+	// more parallel queries, so we break out of the loop.
+	for _, blockIter := range blockIters {
+		// Capture for async query execution below.
+		blockIter := blockIter
 		wg.Add(1)
-		scheduleResult := i.queryWorkersPool.GoWithContext(ctx, func() {
-			startProcessing := time.Now()
-			execBlockFn(ctx, block, query, opts, &state, results, logFields)
-			i.metrics.queryMetrics.blockProcessingTime.RecordDuration(time.Since(startProcessing))
-			wg.Done()
-		})
-		totalWaitTime += scheduleResult.WaitTime
-		if !scheduleResult.Available {
-			state.Lock()
-			state.multiErr = state.multiErr.Add(index.ErrCancelledQuery)
-			state.Unlock()
-			// Did not launch task, need to ensure don't wait for it
+		go func() {
```

Comment:
now we spin up a goroutine per block eagerly and wait for a permit. this allows the goroutine to drive the iterator to completion.

to keep the # of go routines the same with this change, I removed the worker pool and uses the Permits interface to bound the concurrency. the permits impl is the same as the worker pool, but without spawning a goroutine.

Reply:
one change is how all blocks for a query attempt to acquire a permit in parallel. previously a query would sequentially attempt to acquire a worker and then process that worker async. without actually setting a maxResultsPerPermit, this has the potential to make large queries even worse. see the note later about picking a sane default for maxResultsPerPermit.

Reply:
so the assumption here is that we're fine eagerly spinning up O(block) goroutines per query, and having those wait for a permit? Taking this to the extreme, if the system is backed up, we'd have a buildup of goroutines generated by index queries? 

> to keep the # of go routines the same with this change

Not following how this is the same if we're actually generated a goroutine here instead of being capped at O(# of CPUs)

Reply:
yea nvm this is wrong. 

I still think it's ok that we are spinning up O(requests * blocks) goroutines, since it's really going to be dominated by O(requests). If # of requests grows too high, there will be all sorts of other issues in the system.

FWIW the size of backed up requests is at least bounded now by the m3coordinator fetch timeout. 

max go routines = ((retention period hours) / 2 ) * (requests/sec) * fetchTimeout

while that can get kind of high, it's at least bounded.

Reply:
we could limit the # of go routines a single query can spin up to match the theoretically max permits it could acquire. that would at least limit each query to only 8 go routines. basically remove O(blocks) from the equation above.

Reply:
pushed 2fa777b which uses TryAcquire to limit the # of go routines. it's definitely gross, since you have to special case the first permit, but probably worth it.